### PR TITLE
Don't display not relevant groups and groups that don't contain any relevant questions in hierarchy

### LINKED ISSUE #2944: Not relevant repeat groups are visible in the hierarchy
#### Software and hardware versions 
Collect v1.20.0

#### Problem description
Not relevant repeat groups are visible in the hierarchy.

Test form:
[testRepeatCount.xlsx](https://github.com/opendatakit/collect/files/2971649/testRepeatCount.xlsx)
[testRepeatCount.xml.txt](https://github.com/opendatakit/collect/files/2971650/testRepeatCount.xml.txt)

The attached form shows a workaround to hide groups when the number of groups is specified in a form and a user decreased it. More info: https://github.com/opendatakit/collect/issues/868#issuecomment-302466153

The workaround works well but hidden groups are still visible in the hierarchy. If we navigate in the form they are skipped.

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #2944 

#### What has been done to verify that this works as intended?
I tested various forms attached below.

#### Why is this the best possible solution? Were any other approaches considered?
Groups are not marked as not relevant if they contain only not relevant elements so I needed to check all elements inside to decide if a group should be displayed in the hierarchy.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
It should just fix the issue but I'm afraid of any change in that place so @mmarciniak90 again please be super careful and take all the time you need. No rush here because we have discovered many issues in the hierarchy and the code is not easy there.

#### Do we need any specific form for testing your changes? If so, please attach one.
All those forms show the bug:
[hiddenRepeatNetsted.xlsx](https://github.com/opendatakit/collect/files/3038485/hiddenRepeatNetsted.xlsx)
[hiddenRepeat.xlsx](https://github.com/opendatakit/collect/files/3038486/hiddenRepeat.xlsx)
[hidden.xlsx](https://github.com/opendatakit/collect/files/3038487/hidden.xlsx)
[testRepeatCount.xlsx](https://github.com/opendatakit/collect/files/3038503/testRepeatCount.xlsx)

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/logic/FormController.java [hunk 1]
```diff
@@ -938,6 +938,38 @@ private boolean isGroupEmpty() {
...
+                }
+            } else {
+                indices.add(currentChildIndex);
+                currentChildIndex = formEntryModel.incrementIndex(currentChildIndex, false);
+            }
+        }
+        return indices;
+    }
+
+    // A group is relevant if it contains at least one relevant question
```
Comment (cooperka) [inline collect_app/src/main/java/org/odk/collect/android/logic/FormController.java]: Let's make this JavaDoc

File: collect_app/src/main/java/org/odk/collect/android/logic/FormController.java [hunk 2]
```diff
@@ -938,6 +938,38 @@ private boolean isGroupEmpty() {
         return indices;
     }
 
+    private List<FormIndex> getQuestionIndicesForGroup(GroupDef gd) {
```
Comment (cooperka) [inline collect_app/src/main/java/org/odk/collect/android/logic/FormController.java]: It seems like this new method is basically identical to the above `getIndicesForGroup` except that `getEvent(currentChildIndex)` can be `EVENT_REPEAT` also. Let's combine them, maybe something like `getIndicesForGroup(group, eventTypes)`?

**Review by cooperka — COMMENTED** [review]
Seems to work, just some cleanup suggestions 👍 

Unrelated to these changes: I find it strange in general that repeat groups with 0 instances don't appear on the hierarchy. It seems like there should be an item you can navigate to from the hierarchy that will prompt to add an instance. With your `hiddenRepeat` form for example, the hierarchy is completely empty even though there really are places you can navigate to (namely, the prompt for each repeat group). Is this already documented anywhere that you know of, or should I file an issue?

Also unrelated, bug: if you navigate "back" from the end of the `hiddenRepeat` form it takes you to `q1` instead of `q4`. I'll file this too.

File: collect_app/src/main/java/org/odk/collect/android/logic/FormController.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/logic/FormController.java]: Fixed.

File: collect_app/src/main/java/org/odk/collect/android/logic/FormController.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/logic/FormController.java]: Fixed.

**Comment by grzesiek2010** [comment]
> Unrelated to these changes: I find it strange in general that repeat groups with 0 instances don't appear on the hierarchy. It seems like there should be an item you can navigate to from the hierarchy that will prompt to add an instance.

for me it's ok. If we wanted to display empty repeat groups in the hierarchy then fixing #2944 wouldn't be possible because in the form that is attached there we have a workaround for `repeat_count` (decreasing the number of repetitions). The workaround works in that way that we have a nested group which is relevant or not but the parent repetition is always relevant.

**Comment by grzesiek2010** [comment]
> Also unrelated, bug: if you navigate "back" from the end of the hiddenRepeat form it takes you to q1 instead of q4. I'll file this too.

could you provide clear steps?

**Comment by grzesiek2010** [comment]
> Also unrelated, bug: if you navigate "back" from the end of the hiddenRepeat form it takes you to q1 instead of q4. I'll file this too.
could you provide clear steps?

I think I understand it works in that way because the entire second group is not relevant so it's ok. I think that we shouldn't ask a user to add a new group if that group is not relevant (but it's another issue). Look at the form:
image

**Comment by cooperka** [comment]
Ah I see, so it's just skipping all the way back to the beginning because it doesn't find anything relevant. I'll file an issue for just the part about not asking to add a not-relevant group.

**Review by cooperka — APPROVED** [review]
LGTM

**Comment by mmarciniak90** [comment]
@lognaturel 
I'm not sure about this fix. Hierarchy view looks the same if I add many groups and if I do not add any group. Adding even empty groups is nowhere visible. 
Probably it is ok but I want to inform about this behavior.

**Comment by grzesiek2010** [comment]
@lognaturel yeah please share your thoughts. Maybe it's kind of strange because empty groups were visible in the hierarchy before but there is no other way to solve the issue #2944 the entire workaround for `repeat_count` works in that way that we have an additional group inside repeats which is relevant or not as I said above.
So ok maybe it's strange but on the other hand, if you navigate through the form empty groups are skipped so why they are visible in the hierarchy? 
Moreover, I think it's a rare case to have completely empty groups so it's better to fix it in that way than have the issue #2944

**Review by lognaturel — APPROVED** [review]

**Comment by lognaturel** [comment]
I think this change makes a lot of sense: if there is no relevant question in a group, it won't show up in the form filling view so it feels more consistent not to show it in the hierarchy.

Using the same logic, the form filling view gives the user an opportunity to add a repeat instance even when none exist so it would be more consistent for the hierarchy view to show something. I agree with @cooperka that this is a separate question, though.

**Comment by mmarciniak90** [comment]
@lognaturel 
Verified on Android: 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

Verified forms:
- hiddenRepeatNetsted
- hiddenRepeat
- hidden
- testRepeatCount

I confirm that behavior is the same on all Android versions
Reduced number of repeats in testRepeatCount form is correctly represented in the hierarchy view.

|with fix | without fix|
|---|---|
|Screenshot_20190409-113200|Screenshot_20190409-113153|

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Comment by mmarciniak90** [comment]
@grzesiek2010 I noticed one unexpected behavior
1. I added a few groups in hiddenRepeatNetsted form
2. Hierarchy view was empty
3. I saved this form
4. I opened this saved form from `Edit Saved Form` view
5. q1 Repeat Group was visible in the hierarchy view

@opendatakit-bot unlabel "behavior verified"

**Comment by grzesiek2010** [comment]
@mmarciniak90 I think that I fixed the issue. Please continue testing.

**Comment by mmarciniak90** [comment]
Thanks, @grzesiek2010 for a quick fix! 

I checked the mentioned forms again.
I opened them from Fill Blank Form view and Edit Saved Form view.
Now everything looks good.

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Pull request merged**